### PR TITLE
Support pileup flatmapping over any number of RDDs

### LINKED ISSUE #317: support any number of RDDs in `windowTaskFlatMapMultipleRDDs`
see: https://github.com/hammerlab/guacamole/blob/master/src/main/scala/org/hammerlab/guacamole/DistributedUtil.scala#L494

we'll need this for multisample joint somatic calling

### LINKED ISSUE #321: incorrect comment / can't run single unit test
We say:
```
// You can use a line like the following to turn on only a particular unit test for debugging.
  // TestUtil.runOnly = "partitionLociByApproximateReadDepth chr20 synth1 subset"
```
here:
https://github.com/hammerlab/guacamole/blob/master/src/test/scala/org/hammerlab/guacamole/DistributedUtilSuite.scala#L35
and also: https://github.com/hammerlab/guacamole/blob/master/src/test/scala/org/hammerlab/guacamole/util/TestUtil.scala#L46

however, setting neither of these for me limits the tests to a single case.

We should either fix this or delete those lines.

### PR DESCRIPTION
**Pull request by timodonnell**
Generalizes `windowTaskFlatMapMultipleRDDs` to handle any number of RDDs.

Closes #317. Also closes #321.

[Review on Reviewable](https://reviewable.io/reviews/hammerlab/guacamole/324)

### CODE REVIEW

**Comment by timodonnell** [comment]
This changes the code path for all the variations of `pileupFlatMap`, which is performance-sensitive code. I've run some basic tests on the cluster and things seem ok. But we should keep an eye out for any slowdowns or other unexpected effects of this change.

File: src/main/scala/org/hammerlab/guacamole/windowing/SplitIterator.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+   * i'th element in the sequence is an iterator over the values v where k = i.
+   *
+   * @param num number of keys. Each key k must be 0 <= k < num.
+   * @param iterator source iterator of (key, value) pairs. It will be consumed as the result iterators are read.
+   * @tparam V
+   * @return a buffered iterator for each key in 0 .. num.
+   */
+  def split[V](num: Int, iterator: Iterator[(Int, V)]): Seq[BufferedIterator[V]] = {
+    val buffers = (0 until num).map(_ => new collection.mutable.Queue[V])
+    (0 until num).map(index => new SplitIterator[V](index, buffers, iterator))
```
Comment (arahuja) [inline src/main/scala/org/hammerlab/guacamole/windowing/SplitIterator.scala]: Actually, I think this same functionality is possible with `.partition` on an Iterator?

```scala
var otherIterator = iterator
(0 until num).map(index => { 
       indexIterator, otherIterator = otherIterator.partition(_._1 == index)
       indexIterator
})
```

I think this does the same thing, but I did not/run test it.

Comment (timodonnell) [inline src/main/scala/org/hammerlab/guacamole/windowing/SplitIterator.scala]: Yeah, good call. IIRC we used to use `partition` in this way a while ago to split out each reads for each chromosome into a separate iterator, but the overhead of having to nest a `partition` for each key value was substantial (if there are N keys, each element can get buffered N times).

Comment (timodonnell) [inline src/main/scala/org/hammerlab/guacamole/windowing/SplitIterator.scala]: I think this is the PR that got rid of that pattern, so we have an example of what exactly was slow:

https://github.com/hammerlab/guacamole/commit/f557a3fd37b9808235889e4a4693914e20bb8a4e#diff-525c141f106b5f061f77693e79b5e974L558

File: src/main/scala/org/hammerlab/guacamole/DistributedUtil.scala [hunk 2]
```diff
@@ -323,6 +323,34 @@ object DistributedUtil extends Logging {
...
+      readsRDDs,
+      lociPartitions,
+      skipEmpty,
+      halfWindowSize = 0L,
+      initialState = None,
+      function = (maybePileups: Option[Seq[Pileup]], windows: PerSample[SlidingWindow[MappedRead]]) => {
+        val advancedPileups = maybePileups match {
+          case Some(existingPileups) => {
+            existingPileups.zip(windows).map(
+              pileup_and_window => initOrMovePileup(Some(pileup_and_window._1), pileup_and_window._2))
```
Comment (arahuja) [inline src/main/scala/org/hammerlab/guacamole/DistributedUtil.scala]: `pileup_and_window` style?

Comment (timodonnell) [inline src/main/scala/org/hammerlab/guacamole/DistributedUtil.scala]: good call will fix

**Comment by timodonnell** [comment]
updated to address comment

**Comment by arahuja** [comment]
Sorry for the delay, LGTM

**Comment by timodonnell** [comment]
rebased to master

**Pull request merged**